ARTIFICIAL NEURAL NETWORKS

1. Data Exploration and Preprocessing

●	Begin by loading and exploring the "Alphabets_data.csv" dataset. Summarize its key features such as the number of samples, features, and classes.

In [1]:
import pandas as pd

In [2]:
df=pd.read_csv(r"/content/Alphabets_data.csv")

In [3]:
df

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  \
0          T     2     8      3       5      1     8    13      0      6   
1          I     5    12      3       7      2    10     5      5      4   
2          D     4    11      6       8      6    10     6      2      6   
3          N     7    11      6       6      3     5     9      4      6   
4          G     2     1      3       1      1     8     6      6      6   
...      ...   ...   ...    ...     ...    ...   ...   ...    ...    ...   
19995      D     2     2      3       3      2     7     7      7      6   
19996      C     7    10      8       8      4     4     8      6      9   
19997      T     6     9      6       7      5     6    11      3      7   
19998      S     2     3      4       2      1     8     7      2      6   
19999      A     4     9      6       6      2     9     5      3      1   

       xybar  x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0          6      10       8      0       8      0       8  
1         13       3       9      2       8      4      10  
2         10       3       7      3       7      3       9  
3          4       4      10      6      10      2       8  
4          6       5       9      1       7      5      10  
...      ...     ...     ...    ...     ...    ...     ...  
19995      6       6       4      2       8      3       7  
19996     12       9      13      2       9      3       7  
19997     11       9       5      2      12      2       4  
19998     10       6       8      1       9      5       8  
19999      8       1       8      2       7      2       8  

[20000 rows x 17 columns]

In [4]:
df.shape

(20000, 17)

There are 20000 rows and 17 columns in the dataset.

In [5]:
df.columns

Index(['letter', 'xbox', 'ybox', 'width', 'height', 'onpix', 'xbar', 'ybar',
       'x2bar', 'y2bar', 'xybar', 'x2ybar', 'xy2bar', 'xedge', 'xedgey',
       'yedge', 'yedgex'],
      dtype='object')

In [6]:
df.isnull().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


All the features are numerical except for the target fetaure letter.

In [8]:
df.describe()

xbox          ybox         width       height         onpix  \
count  20000.000000  20000.000000  20000.000000  20000.00000  20000.000000   
mean       4.023550      7.035500      5.121850      5.37245      3.505850   
std        1.913212      3.304555      2.014573      2.26139      2.190458   
min        0.000000      0.000000      0.000000      0.00000      0.000000   
25%        3.000000      5.000000      4.000000      4.00000      2.000000   
50%        4.000000      7.000000      5.000000      6.00000      3.000000   
75%        5.000000      9.000000      6.000000      7.00000      5.000000   
max       15.000000     15.000000     15.000000     15.00000     15.000000   

               xbar          ybar         x2bar         y2bar         xybar  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.897600      7.500450      4.628600      5.178650      8.282050   
std        2.026035      2.325354      2.699968      2.380823      2.488475   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.000000      6.000000      3.000000      4.000000      7.000000   
50%        7.000000      7.000000      4.000000      5.000000      8.000000   
75%        8.000000      9.000000      6.000000      7.000000     10.000000   
max       15.000000     15.000000     15.000000     15.000000     15.000000   

            x2ybar        xy2bar         xedge        xedgey         yedge  \
count  20000.00000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.45400      7.929000      3.046100      8.338850      3.691750   
std        2.63107      2.080619      2.332541      1.546722      2.567073   
min        0.00000      0.000000      0.000000      0.000000      0.000000   
25%        5.00000      7.000000      1.000000      8.000000      2.000000   
50%        6.00000      8.000000      3.000000      8.000000      3.000000   
75%        8.00000      9.000000      4.000000      9.000000      5.000000   
max       15.00000     15.000000     15.000000     15.000000     15.000000   

            yedgex  
count  20000.00000  
mean       7.80120  
std        1.61747  
min        0.00000  
25%        7.00000  
50%        8.00000  
75%        9.00000  
max       15.00000

xbox, ybox, width, height  means are around 4–7.

onpix averages around 3.5.

xedge, yedge have lower means(~3), while others like xedgey, yedgex are higher (~8).

Minimum and maximum value for all features is 0 and 15 indicating presence of outliers at the higher end.

Many features appear approximately normal while
some show slight skewness .


In [9]:
df['letter'].nunique()

26

There are 26 unique classes(letters) in the letter column

●	Execute necessary data preprocessing steps including data normalization, managing missing values.

In [10]:
df.isnull().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

There are no missing values present so no imputation is required.

In [11]:
numerical_columns=df.select_dtypes('int').columns

In [12]:
numerical_columns

Index(['xbox', 'ybox', 'width', 'height', 'onpix', 'xbar', 'ybar', 'x2bar',
       'y2bar', 'xybar', 'x2ybar', 'xy2bar', 'xedge', 'xedgey', 'yedge',
       'yedgex'],
      dtype='object')

Normalization:

In [13]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
scaled = scaler.fit_transform(df[numerical_columns])
df[numerical_columns] = scaled

Encoding:

In [14]:
x=df.drop('letter',axis=1)

In [15]:
y=df['letter']

In [16]:
from sklearn.preprocessing import LabelEncoder

In [17]:
le=LabelEncoder()

In [18]:
y_encoded=le.fit_transform(y)

2. Model Implementation

●	Construct a basic ANN model using your chosen high-level neural network library. Ensure your model includes at least one hidden layer.

●	Divide the dataset into training and test sets.

In [63]:
from sklearn.preprocessing import LabelEncoder
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense


# One-hot encode target
y_onehot = to_categorical(y_encoded)

# Features
X = scaled

# Train-test split
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, y_onehot, test_size=0.2, random_state=42)

# Build model for multi-class
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=(X.shape[1],)))
model.add(Dense(32, activation='relu'))
model.add(Dense(len(le.classes_), activation='softmax'))

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train model
history = model.fit(x_train, y_train, epochs=50, batch_size=16, validation_split=0.2)

# Evaluate on test set
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Test Accuracy: {test_accuracy*100:.2f}%")


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.1813 - loss: 2.9284 - val_accuracy: 0.5153 - val_loss: 1.7552
Epoch 2/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.5408 - loss: 1.6109 - val_accuracy: 0.6075 - val_loss: 1.4290
Epoch 3/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.6174 - loss: 1.3543 - val_accuracy: 0.6381 - val_loss: 1.3178
Epoch 4/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.6514 - loss: 1.2288 - val_accuracy: 0.6778 - val_loss: 1.1957
Epoch 5/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.6935 - loss: 1.1275 - val_accuracy: 0.6888 - val_loss: 1.1215
Epoch 6/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7118 - loss: 1.0466 - val_accuracy: 0.7122 - val_loss: 1.0500
Epoch 7/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.7149 - loss: 1.0220 - val_accuracy: 0.7394 - val_loss: 0.9941
Epoch 8/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.7435 - loss: 0.9412 - val_accuracy: 0.

3. Hyperparameter Tuning

●	Modify various hyperparameters, such as the number of hidden layers, neurons per hidden layer, activation functions, and learning rate, to observe their impact on model performance.

In [70]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import RandomizedSearchCV

# Model builder function with **kwargs to accept hyperparameters
def create_model(num_layers=2, neurons=32, activation='relu', learning_rate=0.001):
    model = Sequential()
    model.add(Dense(neurons, activation=activation, input_shape=(X.shape[1],)))
    for _ in range(num_layers - 1):
        model.add(Dense(neurons, activation=activation))
    model.add(Dense(len(le.classes_), activation='softmax'))
    optimizer = Adam(learning_rate=learning_rate)
    model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
    return model

# Wrap model with KerasClassifier and declare hyperparameters in constructor
model = KerasClassifier(
    model=create_model,
    epochs=20,
    verbose=0,
    # Default values for hyperparameters to be tuned
    num_layers=2,
    neurons=32,
    activation='relu',
    learning_rate=0.001,
)

# Hyperparameter distribution for RandomizedSearchCV
param_dist = {
    'num_layers': [1, 2, 3],
    'neurons': [16, 32, 64, 128],
    'activation': ['relu', 'tanh'],
    'learning_rate': [0.001, 0.005, 0.01],
    'batch_size': [16, 32, 64],
}

random_search = RandomizedSearchCV(
    estimator=model,
    param_distributions=param_dist,
    n_iter=10,
    cv=3,
    verbose=2,
    random_state=42,
)

# Fit random search
random_search_result = random_search.fit(x_train, y_train)

print(f"Best Accuracy: {random_search_result.best_score_ * 100:.2f}%")
print("Best Hyperparameters:", random_search_result.best_params_)


Fitting 3 folds for each of 10 candidates, totalling 30 fits


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.005, neurons=128, num_layers=1; total time=  19.3s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.005, neurons=128, num_layers=1; total time=  11.9s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.005, neurons=128, num_layers=1; total time=  11.8s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.01, neurons=128, num_layers=1; total time=  12.7s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.01, neurons=128, num_layers=1; total time=  11.6s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.01, neurons=128, num_layers=1; total time=  12.1s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=16, learning_rate=0.01, neurons=64, num_layers=1; total time=  28.4s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=16, learning_rate=0.01, neurons=64, num_layers=1; total time=  33.4s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=16, learning_rate=0.01, neurons=64, num_layers=1; total time=  29.1s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=32, learning_rate=0.01, neurons=128, num_layers=1; total time=  19.0s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=32, learning_rate=0.01, neurons=128, num_layers=1; total time=  18.7s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=32, learning_rate=0.01, neurons=128, num_layers=1; total time=  18.3s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=relu, batch_size=16, learning_rate=0.005, neurons=32, num_layers=1; total time=  30.1s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=relu, batch_size=16, learning_rate=0.005, neurons=32, num_layers=1; total time=  30.4s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=relu, batch_size=16, learning_rate=0.005, neurons=32, num_layers=1; total time=  32.1s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=16, learning_rate=0.001, neurons=32, num_layers=1; total time=  30.5s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=16, learning_rate=0.001, neurons=32, num_layers=1; total time=  31.5s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=16, learning_rate=0.001, neurons=32, num_layers=1; total time=  30.4s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.001, neurons=16, num_layers=3; total time=  12.3s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.001, neurons=16, num_layers=3; total time=  12.4s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.001, neurons=16, num_layers=3; total time=  12.3s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=relu, batch_size=64, learning_rate=0.001, neurons=16, num_layers=2; total time=  10.8s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=relu, batch_size=64, learning_rate=0.001, neurons=16, num_layers=2; total time=  11.5s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=relu, batch_size=64, learning_rate=0.001, neurons=16, num_layers=2; total time=  11.2s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.01, neurons=16, num_layers=2; total time=  11.0s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.01, neurons=16, num_layers=2; total time=  11.6s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=64, learning_rate=0.01, neurons=16, num_layers=2; total time=  11.5s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=16, learning_rate=0.01, neurons=64, num_layers=2; total time=  35.5s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=16, learning_rate=0.01, neurons=64, num_layers=2; total time=  34.3s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


[CV] END activation=tanh, batch_size=16, learning_rate=0.01, neurons=64, num_layers=2; total time=  31.1s


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Best Accuracy: 85.88%
Best Hyperparameters: {'num_layers': 2, 'neurons': 64, 'learning_rate': 0.01, 'batch_size': 16, 'activation': 'tanh'}


In [71]:
# Final model with best hyperparameters
final_model = create_model(num_layers=2, neurons=64, activation='tanh', learning_rate=0.01)
history = final_model.fit(x_train, y_train, epochs=50, batch_size=16, validation_split=0.2)

test_loss, test_accuracy = final_model.evaluate(x_test, y_test)
print(f"Final Test Accuracy: {test_accuracy*100:.2f}%")


Epoch 1/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.4524 - loss: 1.8298 - val_accuracy: 0.7072 - val_loss: 0.9668
Epoch 2/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7251 - loss: 0.8773 - val_accuracy: 0.7800 - val_loss: 0.7035
Epoch 3/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7792 - loss: 0.6987 - val_accuracy: 0.7984 - val_loss: 0.6444
Epoch 4/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8093 - loss: 0.6031 - val_accuracy: 0.8372 - val_loss: 0.5270
Epoch 5/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8340 - loss: 0.5110 - val_accuracy: 0.8103 - val_loss: 0.5987
Epoch 6/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.8267 - loss: 0.5416 - val_accuracy: 0.8250 - val_loss: 0.5567
Epoch 7/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8582 - loss: 0.4500 - val_accuracy: 0.8572 - val_loss: 0.4593
Epoch 8/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8502 - loss: 0.4681 - val_accuracy: 0.

4. Evaluation

●	Employ suitable metrics such as accuracy, precision, recall, and F1-score to evaluate your model's performance.

In [72]:
from sklearn.metrics import classification_report, accuracy_score

# Model predictions on test data
y_pred_probs = final_model.predict(x_test)
y_pred = y_pred_probs.argmax(axis=1)   # predicted class indices

y_true = y_test.argmax(axis=1)          # true class indices (from one-hot)

# Accuracy
acc = accuracy_score(y_true, y_pred)
print(f"Accuracy: {acc*100:.2f}%\n")

# Classification report (precision, recall, f1-score for each class + averages)
report = classification_report(y_true, y_pred, target_names=le.classes_)
print("Classification Report:\n", report)


125/125 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step
Accuracy: 87.45%

Classification Report:
               precision    recall  f1-score   support

           A       0.88      0.96      0.92       149
           B       0.80      0.79      0.80       153
           C       0.88      0.89      0.89       137
           D       0.64      0.97      0.77       156
           E       0.88      0.81      0.84       141
           F       0.95      0.77      0.85       140
           G       0.82      0.86      0.84       160
           H       0.86      0.60      0.70       144
           I       0.92      0.92      0.92       146
           J       0.96      0.86      0.90       149
           K       0.79      0.84      0.81       130
           L       0.90      0.94      0.92       155
           M       0.90      0.99      0.94       168
           N       0.86      0.93      0.90       151
           O       0.83      0.88      0.85       145
           P       0.89      0.96      0.92       17

The model performs well on most classes, with precision, recall, and F1-scores generally above 0.80.

Some classes, such as 'H', show lower recall, suggesting the model missed some true samples in these categories, which could be a target for future improvement.

Overall, the report shows the model has a balanced performance across most classes, with only minor variations.

●	Discuss the performance differences between the model with default hyperparameters and the tuned model, emphasizing the effects of hyperparameter tuning.

The default model achieved an accuracy of approximately 86.7%, while the hyperparameter-tuned model improved slightly to around 87.4%.
Although the increase in accuracy seems modest, tuning hyperparameters such as the number of layers, neurons, activation function, learning rate, and batch size helped the model converge faster and generalize better on unseen data.
Overall, hyperparameter tuning enhanced the model’s robustness and slightly boosted its predictive accuracy.


